zebrad/src/components/inbound.rs

```diff
@@ -115,85 +145,113 @@ impl Service<zn::Request> for Inbound {
         match req {
-            zn::Request::Peers => match self.address_book.as_ref() {
-                Some(addrs) => {
+            zn::Request::Peers => {
+                if let Setup::Initialized { address_book, .. } = &self.network {
                     // We could truncate the list to try to not reveal our entire
                     // peer set. But because we don't monitor repeated requests,
                     // this wouldn't actually achieve anything, because a crawler
                     // could just repeatedly query it.
-                    let mut peers = addrs.lock().unwrap().sanitized();
+                    let mut peers = address_book.lock().unwrap().sanitized();
                     const MAX_ADDR: usize = 1000; // bitcoin protocol constant
                     peers.truncate(MAX_ADDR);
                     async { Ok(zn::Response::Peers(peers)) }.boxed()
+                } else {
+                    info!("ignoring `Peers` request from remote peer during network setup");
+                    async { Ok(zn::Response::Nil) }.boxed()
                 }
-                None => async { Err("not ready to serve addresses".into()) }.boxed(),
-            },
+            }
             zn::Request::BlocksByHash(hashes) => {
-                let state = self.state.clone();
-                let requests = futures::stream::iter(
-                    hashes
-                        .into_iter()
-                        .map(|hash| zs::Request::Block(hash.into())),
-                );
-
-                state
-                    .call_all(requests)
-                    .try_filter_map(|rsp| {
-                        futures::future::ready(match rsp {
-                            zs::Response::Block(Some(block)) => Ok(Some(block)),
+                // Correctness:
+                //
+                // We can't use `call_all` here, because it leaks buffer slots:
+                // https://github.com/tower-rs/tower/blob/master/tower/src/util/call_all/common.rs#L112
+                use futures::stream::TryStreamExt;
+                hashes
+                    .into_iter()
+                    .map(|hash| zs::Request::Block(hash.into()))
+                    .map(|request| self.state.clone().oneshot(request))
+                    .collect::<futures::stream::FuturesOrdered<_>>()
+                    .try_filter_map(|response| async move {
+                        Ok(match response {
+                            zs::Response::Block(Some(block)) => Some(block),
                             // `zcashd` ignores missing blocks in GetData responses,
                             // rather than including them in a trailing `NotFound`
                             // message
-                            zs::Response::Block(None) => Ok(None),
+                            zs::Response::Block(None) => None,
                             _ => unreachable!("wrong response from state"),
                         })
                     })
```

Comment:
oooh nice, I didn't even realize this was already more or less how it was implemented. This is the block i was most worried about accidentally introducing a behaviour change with, so seeing the original version here brings me a great deal of peace.